File: CppPractice/CppLeetcode/Leetcode52.cpp
#include<set>
using namespace std;

class Solution {
public:
    int publicN=0; 
    int ans=0;
    set<int> colSet;
    set<int> PlusSet;
    set<int> MinusSet;
    bool isValid(int i,int j)
    {
        if (colSet.find(j)!=colSet.end() || PlusSet.find(i+j)!=PlusSet.end() ||MinusSet.find(i-j)!=MinusSet.end())
        {
            return false;
        }
        else
        {
            return true;
        }
    }
    void dfs(int i)
    {
        if (i==publicN)
        {
            ans++;
        }
        else
        {
            for (int j = 0; j < publicN; j++)
            {
                if (isValid(i,j))
                {
                    colSet.insert(j);
                    PlusSet.insert(i+j);
                    MinusSet.insert(i-j);

                    dfs(i+1);
                    
                    colSet.erase(j);
                    PlusSet.erase(i+j);
                    MinusSet.erase(i-j);
                }
            }
        }
    }
    int totalNQueens(int n)
    {
        publicN = n;
        dfs(0);
        return ans;
    }
};

File: CSharpPractice/LeetcodePractice/Leetcode52.cs
public class Solution52 {
    private int ans = 0;
    private int publicN;

    private HashSet<int> colSet = new HashSet<int>();
    private HashSet<int> PlusSet = new HashSet<int>();
    private HashSet<int> MinusSet = new HashSet<int>();

    private void dfs(int i)
    {
        if (i==publicN)
        {
            ans++;
            return;
        }
        else
        {
            for (int j = 0; j < publicN; j++)
            {
                if (isValid(i,j))
                {
                    colSet.Add(j);
                    PlusSet.Add(i+j);
                    MinusSet.Add(i-j);
                    dfs(i+1);
                    colSet.Remove(j);
                    PlusSet.Remove(i+j);
                    MinusSet.Remove(i-j);

                }
                    
            }
 
            
        }
        return;

    }
    private bool isValid(int i,int j)
    {
        if (colSet.Contains(j)|| PlusSet.Contains(i+j)||MinusSet.Contains(i-j))
        {
            return false;
        }
        else
        {
            return true;
        }
    } 
    public int TotalNQueens(int n) 
    {
        publicN = n;
        dfs(0);
        return ans;
        
    }
}